pkg/service/controller/servingcert/secret_updating_controller.go

```diff
@@ -0,0 +1,283 @@
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return sc.secretClient.Secrets(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return sc.secretClient.Secrets(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&kapi.Secret{},
+		resyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    sc.addSecret,
+			UpdateFunc: sc.updateSecret,
+		},
+	)
+	sc.secretHasSynced = sc.secretController.HasSynced
+
+	sc.syncHandler = sc.syncSecret
+
+	return sc
+}
+
+// Run begins watching and syncing.
+func (sc *ServiceServingCertUpdateController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer glog.Infof("Shutting down service signing cert update controller")
+	defer sc.queue.ShutDown()
+
+	go sc.serviceController.Run(stopCh)
+	go sc.secretController.Run(stopCh)
+
+	if !waitForCacheSync(stopCh, sc.serviceHasSynced, sc.secretHasSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+// TODO this is all in the kube library after the 1.5 rebase
+
+// informerSynced is a function that can be used to determine if an informer has synced.  This is useful for determining if caches have synced.
+type informerSynced func() bool
+
+// syncedPollPeriod controls how often you look at the status of your sync funcs
+const syncedPollPeriod = 100 * time.Millisecond
+
+func waitForCacheSync(stopCh <-chan struct{}, cacheSyncs ...informerSynced) bool {
+	err := wait.PollUntil(syncedPollPeriod,
+		func() (bool, error) {
+			for _, syncFunc := range cacheSyncs {
+				if !syncFunc() {
+					return false, nil
+				}
+			}
+			return true, nil
+		},
+		stopCh)
+	if err != nil {
+		glog.V(2).Infof("stop requested")
+		return false
+	}
+
+	glog.V(4).Infof("caches populated")
+	return true
+}
+
+func (sc *ServiceServingCertUpdateController) enqueueSecret(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	sc.queue.Add(key)
+}
+
+func (sc *ServiceServingCertUpdateController) addSecret(obj interface{}) {
+	secret := obj.(*kapi.Secret)
+	if len(secret.Annotations[ServiceNameAnnotation]) == 0 {
+		return
+	}
+
+	glog.V(4).Infof("adding %s", secret.Name)
+	sc.enqueueSecret(secret)
+}
+
+func (sc *ServiceServingCertUpdateController) updateSecret(old, cur interface{}) {
+	secret := cur.(*kapi.Secret)
+	if len(secret.Annotations[ServiceNameAnnotation]) == 0 {
+		// if the current doesn't have a service name, check the old
+		secret = old.(*kapi.Secret)
+		if len(secret.Annotations[ServiceNameAnnotation]) == 0 {
+			return
+		}
+	}
+
+	glog.V(4).Infof("updating %s", secret.Name)
+	sc.enqueueSecret(secret)
+}
+
+func (sc *ServiceServingCertUpdateController) runWorker() {
+	for sc.processNextWorkItem() {
```

Comment:
I must admit I dislike this new form, but that doesn't matter here 😉 I find the body here being much more readable than almost empty for-loop with almost empty method calling another method.